framework/quota/src/org/apache/cloudstack/quota/dao/QuotaBalanceDaoImpl.java

```diff
@@ -0,0 +1,214 @@
+        }
+        return quotaBalances;
+    }
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public List<QuotaBalanceVO> findQuotaBalance(final Long accountId, final Long domainId, final Date startDate, final Date endDate) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+
+        List<QuotaBalanceVO> quotaUsageRecords = null;
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+            if (accountId != null) {
+                sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+            }
+            if (domainId != null) {
+                sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+            }
+            if ((startDate != null) && (endDate != null) && startDate.before(endDate)) {
+                sc.addAnd("updatedOn", SearchCriteria.Op.BETWEEN, startDate, endDate);
+            } else {
+                return new ArrayList<QuotaBalanceVO>();
+            }
+            quotaUsageRecords = listBy(sc);
+            if (quotaUsageRecords.size() == 0) {
+                quotaUsageRecords.addAll(lastQuotaBalanceVO(accountId, domainId, startDate));
+            }
+        } catch (Exception e) {
+            throw new CloudRuntimeException("Unable to find quota balance");
+        } finally {
+            TransactionLegacy.open(opendb).close();
+        }
+
+        return quotaUsageRecords;
+    }
+
+
+    @SuppressWarnings("deprecation")
+    @Override
+    public List<QuotaBalanceVO> lastQuotaBalanceVO(final Long accountId, final Long domainId, final Date pivotDate) {
+        final short opendb = TransactionLegacy.currentTxn().getDatabaseId();
+        List<QuotaBalanceVO> quotaUsageRecords = null;
+        List<QuotaBalanceVO> trimmedRecords = new ArrayList<QuotaBalanceVO>();
+        try (TransactionLegacy txn = TransactionLegacy.open(TransactionLegacy.USAGE_DB)) {
+            Filter filter = new Filter(QuotaBalanceVO.class, "updatedOn", false, 0L, 100L);
+            // ASSUMPTION there will be less than 100 continuous credit
+            // transactions
+            SearchCriteria<QuotaBalanceVO> sc = createSearchCriteria();
+            if (accountId != null) {
+                sc.addAnd("accountId", SearchCriteria.Op.EQ, accountId);
+            }
+            if (domainId != null) {
+                sc.addAnd("domainId", SearchCriteria.Op.EQ, domainId);
+            }
+            if ((pivotDate != null)) {
+                sc.addAnd("updatedOn", SearchCriteria.Op.LTEQ, pivotDate);
+            }
+            quotaUsageRecords = search(sc, filter);
+
+            // get records before startDate to find start balance
+            for (QuotaBalanceVO entry : quotaUsageRecords) {
+                s_logger.info("findQuotaBalance Date=" + entry.getUpdatedOn().toGMTString() + " balance=" + entry.getCreditBalance() + " credit=" + entry.getCreditsId());
```

Comment:
The content of this log message seems more appropriate for ``DEBUG`` than ``INFO``?  Either change the context to be ``INFO`` oriented or reduce the level to ``DEBUG``.